internal/extension/reconciler.go

```diff
@@ -4,12 +4,97 @@ import (
 	"context"
 	"sync"
 	"sync/atomic"
+	"time"
 
 	"github.com/kuadrant/policy-machinery/controller"
 	"github.com/kuadrant/policy-machinery/machinery"
 )
 
-var DAG *atomic.Pointer[StateAwareDAG]
+// nilGuardedPointer is an atomic pointer that provides blocking behavior
+// until the pointer is set to a non-nil value.
+type nilGuardedPointer[T any] struct {
+	ptr  atomic.Pointer[T]
+	mu   sync.Mutex
+	cond *sync.Cond
+}
+
+// newNilGuardedPointer creates a new nilGuardedPointer.
+func newNilGuardedPointer[T any]() *nilGuardedPointer[T] {
+	cgp := &nilGuardedPointer[T]{}
+	cgp.cond = sync.NewCond(&cgp.mu)
+	return cgp
+}
+
+// set sets the pointer to x and signals any goroutines waiting for a non-nil value.
+func (c *nilGuardedPointer[T]) set(x T) {
+	c.ptr.Store(&x)
+
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	c.cond.Broadcast()
+}
+
+// get returns the current value of the pointer without blocking.
+func (c *nilGuardedPointer[T]) get() *T {
+	return c.ptr.Load()
+}
+
+// getWait blocks until the pointer is set to a non-nil value and then returns that value.
+func (c *nilGuardedPointer[T]) getWait() T {
+	// First try a quick non-blocking check
+	if val := c.ptr.Load(); val != nil {
+		return *c.ptr.Load()
+	}
+
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	for c.ptr.Load() == nil {
+		c.cond.Wait()
+	}
+
+	return *c.ptr.Load()
+}
+
+// getWaitWithTimeout blocks until the pointer is set to a non-nil value or until the timeout is reached.
+// Returns the current value of the pointer and a boolean indicating whether the value was set before the timeout.
+func (c *nilGuardedPointer[T]) getWaitWithTimeout(timeout time.Duration) (*T, bool) {
+	// First try a quick non-blocking check
+	if val := c.ptr.Load(); val != nil {
+		return c.ptr.Load(), true
+	}
+
+	timer := time.NewTimer(timeout)
+	defer timer.Stop()
+
+	done := make(chan struct{})
+	var result *T
+	var success bool
+
+	go func() {
+		c.mu.Lock()
+		defer c.mu.Unlock()
+
+		for c.ptr.Load() == nil {
+			c.cond.Wait()
+		}
+
+		result = c.ptr.Load()
+		success = true
+		close(done)
+	}()
+
+	select {
+	case <-done:
+		return result, success
+	case <-timer.C:
+		return c.ptr.Load(), false
+	}
```

Comment:
Is there a similar problem here or am I misunderstanding? maybe something like this fixes it?

```suggestion
	result := make(chan *T, 1)

	go func() {
		c.mu.Lock()
		defer c.mu.Unlock()

		for c.ptr.Load() == nil {
			c.cond.Wait()
		}

		val := c.ptr.Load()
		result <- val
	}()

	select {
	case val := <-result:
		return val, true
	case <-timer.C:
		return c.ptr.Load(), false
	}
```

Reply:
as I read it, same code indeed...